### ISSUE TITLE
longhorn-manager pod is crashing with CrashLoopBackOff   

### ISSUE THREAD

**Issue by hirendave47** [longhorn/longhorn#4462]
Hello,
I am trying to deploy longhorn on a Kubernetes cluster in a vLAB environment.
But the longhorn-manager pod is failing.

Packages used:
kubelet v1.24.3-00
containerd v1.24.3-00
kubectl v1.24.3-00
kubeadm v1.24.3-00
Longhorn v1.3.1

Node OS: Ubuntu 20.04.3 LTS
Internet connectivity using a local Squid proxy server

`# kubectl get all
NAME                                               READY   STATUS             RESTARTS        AGE
pod/longhorn-admission-webhook-8498c969d4-9992g    1/1     Running            0               24m
pod/longhorn-admission-webhook-8498c969d4-mlxx2    1/1     Running            0               24m
pod/longhorn-conversion-webhook-78fc4df57c-b92b5   1/1     Running            0               24m
pod/longhorn-conversion-webhook-78fc4df57c-gmntz   1/1     Running            0               24m
pod/longhorn-driver-deployer-855dfc74ff-tnxpq      0/1     Init:0/1           0               24m
pod/longhorn-manager-58lsq                         0/1     CrashLoopBackOff   9 (3m33s ago)   24m
pod/longhorn-manager-687ds                         0/1     CrashLoopBackOff   9 (3m28s ago)   24m
pod/longhorn-manager-j9nsm                         0/1     CrashLoopBackOff   9 (3m21s ago)   24m
pod/longhorn-ui-58b8cc8d8b-dg5rh                   1/1     Running            0               24m

NAME                                  TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
service/longhorn-admission-webhook    ClusterIP   10.99.165.112   <none>        9443/TCP   24m
service/longhorn-backend              ClusterIP   10.102.158.48   <none>        9500/TCP   24m
service/longhorn-conversion-webhook   ClusterIP   10.101.199.31   <none>        9443/TCP   24m
service/longhorn-engine-manager       ClusterIP   None            <none>        <none>     24m
service/longhorn-frontend             ClusterIP   10.99.27.208    <none>        80/TCP     24m
service/longhorn-replica-manager      ClusterIP   None            <none>        <none>     24m

NAME                              DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
daemonset.apps/longhorn-manager   3         3         0       3            0           <none>          24m

NAME                                          READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/longhorn-admission-webhook    2/2     2            2           24m
deployment.apps/longhorn-conversion-webhook   2/2     2            2           24m
deployment.apps/longhorn-driver-deployer      0/1     1            0           24m
deployment.apps/longhorn-ui                   1/1     1            1           24m

NAME                                                     DESIRED   CURRENT   READY   AGE
replicaset.apps/longhorn-admission-webhook-8498c969d4    2         2         2       24m
replicaset.apps/longhorn-conversion-webhook-78fc4df57c   2         2         2       24m
replicaset.apps/longhorn-driver-deployer-855dfc74ff      1         1         0       24m
replicaset.apps/longhorn-ui-58b8cc8d8b                   1         1         1       24m
#
# kubectl describe pod/longhorn-manager-58lsq
Name:         longhorn-manager-58lsq
Namespace:    longhorn-system
Priority:     0
Node:         node1/192.168.239.21
Start Time:   Mon, 22 Aug 2022 14:44:50 +0000
Labels:       app=longhorn-manager
              app.kubernetes.io/instance=longhorn
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=longhorn
              app.kubernetes.io/version=v1.3.1
              controller-revision-hash=77b984974f
              helm.sh/chart=longhorn-1.3.1
              pod-template-generation=1
Annotations:  cni.projectcalico.org/containerID: f8b8b98e75632307376a279b30650b238c0fa8730563acefb359ea81e236b143
              cni.projectcalico.org/podIP: 172.16.229.89/32
              cni.projectcalico.org/podIPs: 172.16.229.89/32
Status:       Running
IP:           172.16.229.89
IPs:
  IP:           172.16.229.89
Controlled By:  DaemonSet/longhorn-manager
Init Containers:
  wait-longhorn-admission-webhook:
    Container ID:  containerd://cf538d487b84c0185891c285ae78de317772e21337d0ce870b750265ba002e0a
    Image:         longhornio/longhorn-manager:v1.3.1
    Image ID:      docker.io/longhornio/longhorn-manager@sha256:8cb9792ada5b4a2b8ff8b67b878b654c25d0318a81b07c01e786c753d4c46523
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
      while [ $(curl -m 1 -s -o /dev/null -w "%{http_code}" -k https://longhorn-admission-webhook:9443/v1/healthz) != "200" ]; do echo waiting; sleep 2; done
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Mon, 22 Aug 2022 14:44:51 +0000
      Finished:     Mon, 22 Aug 2022 14:44:58 +0000
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-fsvp9 (ro)
Containers:
  longhorn-manager:
    Container ID:  containerd://922de27020a84289b3ecbc24ac5ec766e2147b81a77d630b06b7aba7aac42121
    Image:         longhornio/longhorn-manager:v1.3.1
    Image ID:      docker.io/longhornio/longhorn-manager@sha256:8cb9792ada5b4a2b8ff8b67b878b654c25d0318a81b07c01e786c753d4c46523
    Port:          9500/TCP
    Host Port:     0/TCP
    Command:
      longhorn-manager
      -d
      daemon
      --engine-image
      longhornio/longhorn-engine:v1.3.1
      --instance-manager-image
      longhornio/longhorn-instance-manager:v1_20220808
      --share-manager-image
      longhornio/longhorn-share-manager:v1_20220808
      --backing-image-manager-image
      longhornio/backing-image-manager:v3_20220808
      --manager-image
      longhornio/longhorn-manager:v1.3.1
      --service-account
      longhorn-service-account
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Mon, 22 Aug 2022 15:06:10 +0000
      Finished:     Mon, 22 Aug 2022 15:06:10 +0000
    Ready:          False
    Restart Count:  9
    Readiness:      tcp-socket :9500 delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      POD_NAMESPACE:  longhorn-system (v1:metadata.namespace)
      POD_IP:          (v1:status.podIP)
      NODE_NAME:       (v1:spec.nodeName)
    Mounts:
      /host/dev/ from dev (rw)
      /host/proc/ from proc (rw)
      /tls-files/ from longhorn-grpc-tls (rw)
      /var/lib/longhorn/ from longhorn (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-fsvp9 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  dev:
    Type:          HostPath (bare host directory volume)
    Path:          /dev/
    HostPathType:
  proc:
    Type:          HostPath (bare host directory volume)
    Path:          /proc/
    HostPathType:
  longhorn:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/longhorn/
    HostPathType:
  longhorn-grpc-tls:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  longhorn-grpc-tls
    Optional:    true
  kube-api-access-fsvp9:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/disk-pressure:NoSchedule op=Exists
                             node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists
                             node.kubernetes.io/pid-pressure:NoSchedule op=Exists
                             node.kubernetes.io/unreachable:NoExecute op=Exists
                             node.kubernetes.io/unschedulable:NoSchedule op=Exists
Events:
  Type     Reason     Age                 From               Message
  ----     ------     ----                ----               -------
  Normal   Scheduled  25m                 default-scheduler  Successfully assigned longhorn-system/longhorn-manager-58lsq to node1
  Normal   Pulled     25m                 kubelet            Container image "longhornio/longhorn-manager:v1.3.1" already present on machine
  Normal   Created    25m                 kubelet            Created container wait-longhorn-admission-webhook
  Normal   Started    25m                 kubelet            Started container wait-longhorn-admission-webhook
  Normal   Created    24m (x4 over 25m)   kubelet            Created container longhorn-manager
  Normal   Started    24m (x4 over 25m)   kubelet            Started container longhorn-manager
  Normal   Pulled     23m (x5 over 25m)   kubelet            Container image "longhornio/longhorn-manager:v1.3.1" already present on machine
  Warning  BackOff    5s (x122 over 25m)  kubelet            Back-off restarting failed container
# kubectl logs pod/longhorn-manager-58lsq
Defaulted container "longhorn-manager" out of: longhorn-manager, wait-longhorn-admission-webhook (init)
W0822 15:06:10.307900       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2022-08-22T15:06:10Z" level=info msg="cannot list the content of the src directory /var/lib/rancher/longhorn/engine-binaries for the copy, will do nothing: Failed to execute: nsenter [--mount=/host/proc/1/ns/mnt --net=/host/proc/1/ns/net bash -c ls /var/lib/rancher/longhorn/engine-binaries/*], output , stderr, ls: cannot access '/var/lib/rancher/longhorn/engine-binaries/*': No such file or directory\n, error exit status 2"
I0822 15:06:10.353730       1 leaderelection.go:248] attempting to acquire leader lease longhorn-system/longhorn-manager-upgrade-lock...
I0822 15:06:10.370533       1 leaderelection.go:258] successfully acquired lease longhorn-system/longhorn-manager-upgrade-lock
time="2022-08-22T15:06:10Z" level=info msg="Start upgrading"
time="2022-08-22T15:06:10Z" level=info msg="setting default-engine-image not found"
time="2022-08-22T15:06:10Z" level=error msg="Upgrade failed: upgrade API version failed: cannot create CRDAPIVersionSetting: Internal error occurred: failed calling webhook \"validator.longhorn.io\": failed to call webhook: Post \"https://longhorn-admission-webhook.longhorn-system.svc:9443/v1/webhook/validaton?timeout=10s\": Forbidden"
time="2022-08-22T15:06:10Z" level=info msg="Upgrade leader lost: node1"
time="2022-08-22T15:06:10Z" level=fatal msg="Error starting manager: upgrade API version failed: cannot create CRDAPIVersionSetting: Internal error occurred: failed calling webhook \"validator.longhorn.io\": failed to call webhook: Post \"https://longhorn-admission-webhook.longhorn-system.svc:9443/v1/webhook/validaton?timeout=10s\": Forbidden"
#`

`# kubectl get all -n kube-system
NAME                                           READY   STATUS    RESTARTS      AGE
pod/calico-kube-controllers-555bc4b957-4r4bh   1/1     Running   2 (36m ago)   3d1h
pod/calico-node-9xlhh                          1/1     Running   2 (36m ago)   3d1h
pod/calico-node-q2hxk                          1/1     Running   2 (35m ago)   3d1h
pod/calico-node-xgp9j                          1/1     Running   2 (34m ago)   3d1h
pod/coredns-6d4b75cb6d-b8rm4                   1/1     Running   2 (36m ago)   3d1h
pod/coredns-6d4b75cb6d-nf7lj                   1/1     Running   2 (36m ago)   3d1h
pod/etcd-node1                            1/1     Running   2 (36m ago)   3d1h
pod/kube-apiserver-node1                  1/1     Running   2 (36m ago)   3d1h
pod/kube-controller-manager-node1         1/1     Running   2 (36m ago)   3d1h
pod/kube-proxy-cw664                           1/1     Running   2 (34m ago)   3d1h
pod/kube-proxy-mhxw7                           1/1     Running   2 (36m ago)   3d1h
pod/kube-proxy-w47rv                           1/1     Running   2 (35m ago)   3d1h
pod/kube-scheduler-node1                  1/1     Running   2 (36m ago)   3d1h

NAME               TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
service/kube-dns   ClusterIP   10.96.0.10   <none>        53/UDP,53/TCP,9153/TCP   3d1h

NAME                         DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
daemonset.apps/calico-node   3         3         3       3            3           kubernetes.io/os=linux   3d1h
daemonset.apps/kube-proxy    3         3         3       3            3           kubernetes.io/os=linux   3d1h

NAME                                      READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/calico-kube-controllers   1/1     1            1           3d1h
deployment.apps/coredns                   2/2     2            2           3d1h

NAME                                                 DESIRED   CURRENT   READY   AGE
replicaset.apps/calico-kube-controllers-555bc4b957   1         1         1       3d1h
replicaset.apps/coredns-6d4b75cb6d                   2         2         2       3d1h
#`

Any help would be highly appreciated.

**Comment by derekbit** [comment]
Can you
1. show the logs of the `longhorn-admission-webhook`?
2. enter the longhorn-conversion-webhook or longhorn-ui, and then check the connectivity to `https://longhorn-admission-webhook.longhorn-system.svc:9443`

You probably can try to re-deploy LH and check if it happend again. 
BTW, does LH v1.2.4 work in the env?

**Comment by hirendave47** [comment]
Thanks Derek for your prompt response, here are the requested logs.
I have already tried to re-deploy the longhorn multiple times but each time it's giving same issue.
The longhorn-admission-webhook logs are attached in a file and below are the connectivity logs.

# kubectl exec -it pod/longhorn-conversion-webhook-78fc4df57c-b92b5 -- /bin/bash
I have no name!@longhorn-conversion-webhook-78fc4df57c-b92b5:/> telnet longhorn-admission-webhook.longhorn-system.svc 9443
Trying 10.99.165.112...
Connected to longhorn-admission-webhook.longhorn-system.svc.
Escape character is '^]'.
^]
telnet> quit
Connection closed.
I have no name!@longhorn-conversion-webhook-78fc4df57c-b92b5:/>
I have no name!@longhorn-conversion-webhook-78fc4df57c-b92b5:/> curl -vI http://longhorn-admission-webhook.longhorn-system.svc:9443/v1/webhook/validaton
*   Trying 10.99.165.112:9443...
* TCP_NODELAY set
* Connected to longhorn-admission-webhook.longhorn-system.svc (10.99.165.112) port 9443 (#0)
> HEAD /v1/webhook/validaton HTTP/1.1
> Host: longhorn-admission-webhook.longhorn-system.svc:9443
> User-Agent: curl/7.66.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
* HTTP 1.0, assume close after body
< HTTP/1.0 400 Bad Request
HTTP/1.0 400 Bad Request

<
* Excess found: excess = 48 url = /v1/webhook/validaton (zero-length body)
* Closing connection 0
[longhorn-admission-webhook_logs.txt](https://github.com/longhorn/longhorn/files/9398951/longhorn-admission-webhook_logs.txt)

I have no name!@longhorn-conversion-webhook-78fc4df57c-b92b5:/> curl -vI https://longhorn-admission-webhook.longhorn-system.svc:9443/v1/webhook/validaton
*   Trying 10.99.165.112:9443...
* TCP_NODELAY set
* Connected to longhorn-admission-webhook.longhorn-system.svc (10.99.165.112) port 9443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS alert, unknown CA (560):
* SSL certificate problem: unable to get local issuer certificate
* Closing connection 0
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
I have no name!@longhorn-conversion-webhook-78fc4df57c-b92b5:/>

**Comment by hirendave47** [comment]
I tried installing Longhorn v1.2.4 and all PODs seems to be running fine but I can still see the same error in the longhorn-manager pod.
[longhorn_v1.2.4_logs.txt](https://github.com/longhorn/longhorn/files/9399118/longhorn_v1.2.4_logs.txt)

**Comment by derekbit** [comment]
@hirendave47 
Thanks for the information.
The connectivity issue looks related to the proxy.
How did you bootstrap your kubernetes cluster? Use kubeadm, right?

**Comment by hirendave47** [comment]
Yes, I have used kubeadm command.
For proxy, I am using a local Ubuntu VM with Squid and on k8s nodes, I have exported below env vars.
export http_proxy=192.168.239.27:3128
export https_proxy=192.168.239.27:3128
export no_proxy=192.168.239.0/24,10.96.0.0/12,127.0.0.1,172.16.0.0/12

This proxy is configured in containerd service as well to download images.

How can I correct this proxy issue ?

**Comment by derekbit** [comment]
Can you try this in v1.3.1
```
export no_proxy=$no_proxy,.svc,.svc.cluster.local
```

**Comment by hirendave47** [comment]
Thanks Derek, I tried your solution, exported env var no_proxy, updated it in containerd service, restarted, killed longhorn-manager pods but same issue. Then I uninstalled LH v1.2.4 and deployed LH v1.3.1 again but still getting same error.

Defaulted container "longhorn-manager" out of: longhorn-manager, wait-longhorn-admission-webhook (init)
W0823 05:08:24.387805       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2022-08-23T05:08:24Z" level=info msg="cannot list the content of the src directory /var/lib/rancher/longhorn/engine-binaries for the copy, will do nothing: Failed to execute: nsenter [--mount=/host/proc/1/ns/mnt --net=/host/proc/1/ns/net bash -c ls /var/lib/rancher/longhorn/engine-binaries/*], output , stderr, ls: cannot access '/var/lib/rancher/longhorn/engine-binaries/*': No such file or directory\n, error exit status 2"
I0823 05:08:24.455788       1 leaderelection.go:248] attempting to acquire leader lease longhorn-system/longhorn-manager-upgrade-lock...
I0823 05:08:24.466156       1 leaderelection.go:258] successfully acquired lease longhorn-system/longhorn-manager-upgrade-lock
time="2022-08-23T05:08:24Z" level=info msg="Start upgrading"
time="2022-08-23T05:08:24Z" level=info msg="setting default-engine-image not found"
time="2022-08-23T05:08:24Z" level=error msg="Upgrade failed: upgrade API version failed: cannot create CRDAPIVersionSetting: Internal error occurred: failed calling webhook \"validator.longhorn.io\": failed to call webhook: Post \"https://longhorn-admission-webhook.longhorn-system.svc:9443/v1/webhook/validaton?timeout=10s\": Forbidden"
time="2022-08-23T05:08:24Z" level=info msg="Upgrade leader lost: node1"
time="2022-08-23T05:08:24Z" level=fatal msg="Error starting manager: upgrade API version failed: cannot create CRDAPIVersionSetting: Internal error occurred: failed calling webhook \"validator.longhorn.io\": failed to call webhook: Post \"https://longhorn-admission-webhook.longhorn-system.svc:9443/v1/webhook/validaton?timeout=10s\": Forbidden"

**Comment by derekbit** [comment]
@hirendave47 
Got it. I need to futher investigate this issue.
But we don't introduce the webhook in v1.2.4. The [error message ](https://github.com/longhorn/longhorn/issues/4462#issuecomment-1223486308) looks quite normal, so you can try v1.2.4 and create pod, pvc and etc like the [examples](https://github.com/longhorn/longhorn/tree/master/examples) in longhorn.

**Comment by hirendave47** [comment]
I did one more experiment, I removed proxy configuration from every node and re-created the LH but no luck.

**Comment by derekbit** [comment]
@hirendave47 
On more question, how did you update no_proxy in your containerd service?

**Comment by hirendave47** [comment]
I am using environment variable as per below.

$ vi /lib/systemd/system/containerd.service
***output_truncated***
Delegate=yes
KillMode=process
Restart=always
Environment="HTTP_PROXY=http://192.168.239.27:3128/"
Environment="NO_PROXY=192.168.239.0/24,10.96.0.0/12,127.0.0.1"
***output_truncated***
$ systemctl daemon-reload
$ systemctl restart containerd

In /etc/profile, below variables are added.
export http_proxy=192.168.239.27:3128
export https_proxy=192.168.239.27:3128
export no_proxy="192.168.239.0/24,10.96.0.0/12,127.0.0.1"

**Comment by derekbit** [comment]
So, if you replace the `Environment="NO_PROXY...` with `Environment="NO_PROXY=.svc,.svc.cluster.local,192.168.239.0/24,10.96.0.0/12,127.0.0.1"`, still not work, right?

**Comment by hirendave47** [comment]
yes, I tried it earlier but didn't worked.
I also tested by removing proxy configuration from /etc/profile and containerd.service files completely and rebooted all the nodes but still same error for longhorn-manager pod.

**Comment by hirendave47** [comment]
An update on this issue.... No actual solution yet but found one work-around... able to install latest LH version this way...

$ kubectl get ValidatingWebhookConfiguration -n longhorn-system
NAME                         WEBHOOKS   AGE
longhorn-webhook-validator   1          4m
$ kubectl delete ValidatingWebhookConfiguration longhorn-webhook-validator -n longhorn-system
warning: deleting cluster-scoped resources, not scoped to the provided namespace
validatingwebhookconfiguration.admissionregistration.k8s.io "longhorn-webhook-validator" deleted
$

**Comment by derekbit** [comment]
> An update on this issue.... No actual solution yet but found one work-around... able to install latest LH version this way...
> 
> $ kubectl get ValidatingWebhookConfiguration -n longhorn-system NAME WEBHOOKS AGE longhorn-webhook-validator 1 4m $ kubectl delete ValidatingWebhookConfiguration longhorn-webhook-validator -n longhorn-system warning: deleting cluster-scoped resources, not scoped to the provided namespace validatingwebhookconfiguration.admissionregistration.k8s.io "longhorn-webhook-validator" deleted $

But the admission webhook is an important and a must component in LH system. Removing it will introduce unexpected issues...

**Comment by hirendave47** [comment]
Well, I found the solution... or the bug in my configuration... :)

Previously, I was using a custom dnsDomain which was also configured as search directive in /etc/resolv.conf on the host VM. The resolv.conf file on the host VM was redirecting all such traffic to external DNS server defined via nameserver directive.

*** Not working ****
$ grep  dnsDomain ClusterConfiguration.yaml
  dnsDomain: mydomain.lab
$

*** DNS client configuration on all nodes ***
$ cat /etc/resolv.conf
nameserver 192.168.239.24
search mydomain.lab
$

I changed the dnsDomain directive in cluster bootstrap configuration to default one and Longhorn latest version is working fine after this change.

*** Working fine after changing this to something other then host VM domain ****
$ grep  dnsDomain ClusterConfiguration.yaml-1
  dnsDomain: cluster.local
$

No change in host VM resolv.conf.
